r1cs/gadgets/std/src/groups/nonnative/nonnative_group_gadget.rs

```diff
@@ -0,0 +1,844 @@
+use std::{borrow::Borrow, marker::PhantomData};
+
+#[derive(Derivative)]
+#[derivative(Debug, Clone)]
+pub struct GroupAffineNonNativeGadget<
+    P: SWModelParameters<BaseField = SimulationF>,
+    ConstraintF: PrimeField,
+    SimulationF: PrimeField + SquareRootField
+> {
+    pub x:   NonNativeFieldGadget<SimulationF, ConstraintF>,
+    pub y:   NonNativeFieldGadget<SimulationF, ConstraintF>,
+    pub infinity:   Boolean,
+    _params: PhantomData<P>,
+}
+
+
+impl<P, ConstraintF, SimulationF> GroupGadget<SWProjective<P>, ConstraintF>
+for GroupAffineNonNativeGadget<P, ConstraintF, SimulationF>
+    where
+        P: SWModelParameters<BaseField = SimulationF>,
+        ConstraintF: PrimeField,
+        SimulationF: PrimeField + SquareRootField
+{
+    type Value = SWProjective<P>;
+    type Variable = ();
+
+    fn add<CS: ConstraintSystem<ConstraintF>>(
+        &self,
+        cs: CS,
+        other: &Self,
+    ) -> Result<Self, SynthesisError> {
+        self.add_internal(cs, other, true)
+    }
+
+    #[inline]
+    fn zero<CS: ConstraintSystem<ConstraintF>>(mut cs: CS) -> Result<Self, SynthesisError> {
+        Ok(Self::new(
+            NonNativeFieldGadget::zero(cs.ns(|| "zero"))?,
+            NonNativeFieldGadget::one(cs.ns(|| "one"))?,
+            Boolean::constant(true),
+        ))
+    }
+
+
+    #[inline]
+    fn is_zero<CS: ConstraintSystem<ConstraintF>>(&self, _: CS) -> Result<Boolean, SynthesisError>{
+        Ok(self.infinity)
+    }
+
+    #[inline]
+    fn double_in_place<CS: ConstraintSystem<ConstraintF>>(
+        &mut self,
+        mut cs: CS,
+    ) -> Result<(), SynthesisError> {
+        let a = P::COEFF_A;
+
+        let x_squared = self.x.square(cs.ns(|| "x^2"))?;
+
+        let one = P::BaseField::one();
+        let two = one.double();
+        let three = two + &one;
+
+        let three_x_squared = x_squared.mul_by_constant(cs.ns(|| "3 * x^2"), &three)?;
+        let three_x_squared_plus_a = three_x_squared.add_constant(cs.ns(|| "3 * x^2 + a"), &a)?;
+
+        let two_y = self.y.double(cs.ns(|| "2y"))?;
+
+        let lambda = NonNativeFieldGadget::alloc(cs.ns(|| "lambda"), || {
+            let y_doubled_inv = NonNativeFieldGadget::get_value(&two_y).get()?.inverse().get()?;
+            Ok(three_x_squared_plus_a.get_value().get()? * &y_doubled_inv)
+        })?;
+
+        // Check lambda
+        lambda.mul_equals(cs.ns(|| "check lambda"), &two_y, &three_x_squared_plus_a)?;
+
+        // Allocate fresh x and y as a temporary workaround to reduce the R1CS density.
+        let x = NonNativeFieldGadget::alloc(
+            cs.ns(|| "new x"),
+            || {
+                let lambda_val = lambda.get_value().get()?;
+                let x_val = self.x.get_value().get()?;
+                Ok((lambda_val * &lambda_val) - &x_val - &x_val)
+            }
+        )?;
```

Comment:
Is this really needed, or, did we explicitly measure the R1CS density without that tweak? 
It's the non-native variant, so maybe the non-native sub enforces already enough to keep the R1CS density low.

Reply:
I would say to put a TODO for the moment and address this later (we will use this gadget in context of Groth16 for which density doesn't really matter)